Automattic/newspack-theme
# Add 'Highlight/focus on' menu to theme

### All Submissions:


### Changes proposed in this Pull Request:

This PR adds a new menu to the theme, called 'Highlight Menu', intended to be used to draw traffic to specific posts. 

I made some decisions about how this menu should work based on how the header currently behaves; I've outlined them below. I'd be very interested in feedback on that, as well as the code and how well things actually function. I'm open to feedback on any of this:

* **The menu displays its name in front of it.** This means whatever you name the menu will appear before the links you add to it. This seemed like the easiest way to give it an editable text label without adding a new field.
* **This menu doesn't display on mobile.** Looking at sites with similar menus, including the Washington post, the menu typically was hidden on smaller screens, even from the mobile menu.
* **This menu only displays one line of links.** Any further links are forced out of sight (though they should still be reachable via keyboard navigation, since they're not fully hidden). This was the easiest way to keep things looking tidy; since the text used for menu links is editable, it is possible to give stories/categories shorter text labels if needed. (I admit, this was also borrowed from the WaPo's approach). 
* **This menu does not appear when you use the 'featured image behind' or 'featured image beside' styles.** I mostly made this decision based on keeping those styles as bold and eye catching as possible -- it was tricky to figure out how to work this new menu in without causing a weird gap. 

Closes #441.

(Note: Until #451 is merged, using the behind and beside featured image styles may cause some unrelated visual weirdness).

### How to test the changes in this Pull Request:

1. Apply the PR and run `npm run build`.
2. Navigate to Customizer > Menus, and add a menu to the new "Highlight Menu" location. Start with a longer menu.
3. View the menu on the front end; confirm it displays in a way that seems "correct" ; here it is with the default header and style pack settings, and a menu called "Don't Miss":

![image](https://user-images.githubusercontent.com/177561/66159823-a5943f00-e5dd-11e9-8241-b10ad02a6c00.png)

4. Try shrinking the browser window; confirm that the last menu item drops off, and is no longer visible, when there's not enough space for it:

![image](https://user-images.githubusercontent.com/177561/66159845-ad53e380-e5dd-11e9-95e9-9a62215e9210.png)

5. Try tabbing through the page using the keyboard to navigate; make sure once you get to the end of the visible links in the menu, the ones that are cut off appear and can be tabbed to:

![image](https://user-images.githubusercontent.com/177561/66159872-b9d83c00-e5dd-11e9-934c-6decc9264193.png)

6. Try shrinking down the browser window until the mobile menu is visible, and confirm that the highlight menu is hidden:

![image](https://user-images.githubusercontent.com/177561/66159973-f73cc980-e5dd-11e9-866a-1af7b065546d.png)

7. Navigate to Customize > Header Settings and switch to the centred logo; confirm that this menu is also centred (may be harder to tell at some breakpoints, but if you scale down your screen to make a link drop off, it should be visibly centred):

![image](https://user-images.githubusercontent.com/177561/66161598-84355200-e5e1-11e9-916a-f0d695db7a7e.png)

8. Navigate to Customize > Header Settings and switch to the 'short header'; confirm that the spacing still looks okay:

![image](https://user-images.githubusercontent.com/177561/66163136-7fbe6880-e5e4-11e9-918a-fcb7e9abf6ea.png)

9. Navigate to Customize > Header Settings, and switch to the 'solid background':

![image](https://user-images.githubusercontent.com/177561/66163207-9fee2780-e5e4-11e9-88aa-24cd35e465e8.png)

![image](https://user-images.githubusercontent.com/177561/66163266-bf855000-e5e4-11e9-8be3-8d9ac3fd0c2c.png)

10. Navigate to Customize > Style Packs and switch to 'Style 2' -- it's the only one with a very different header. Navigate to Customizer > Header Settings, and uncheck all of the options:

![image](https://user-images.githubusercontent.com/177561/66163464-44706980-e5e5-11e9-9907-e34e421f413a.png)

11. Navigate to Customizer > Header Settings; check 'solid background':

![image](https://user-images.githubusercontent.com/177561/66163539-679b1900-e5e5-11e9-973c-954264195dbf.png)

12. Navigate to Customize > Header Settings; check 'short header':

![image](https://user-images.githubusercontent.com/177561/66163727-d6787200-e5e5-11e9-9751-4ab14c528a88.png)

13. Create a new post; add a featured image and view it:

![image](https://user-images.githubusercontent.com/177561/66163892-396a0900-e5e6-11e9-9f3d-b433dfa50112.png)

14. Edit the post; select the 'Behind article title' style and confirm the 'highlight menu' is hidden:

![image](https://user-images.githubusercontent.com/177561/66163966-6cac9800-e5e6-11e9-8462-05360f88d9d8.png)

15. Edit the post; select the 'Beside article title'; confirm the 'highlight menu' is hidden:

![image](https://user-images.githubusercontent.com/177561/66164048-9f569080-e5e6-11e9-8b9e-04743044a67b.png)

### Other information:
